clap/codepoint_counting_writer.zig

```diff
@@ -0,0 +1,47 @@
+const std = @import("std");
+
+/// A Writer that counts how many codepoints has been written to it.
+pub fn CodepointCountingWriter(comptime WriterType: type) type {
+    return struct {
+        codepoints_written: u64,
+        child_stream: WriterType,
+
+        pub const Error = WriterType.Error || error{
+            TruncatedInput,
+            Utf8CodepointTooLarge,
+            Utf8EncodesSurrogateHalf,
+            Utf8ExpectedContinuation,
+            Utf8InvalidStartByte,
+            Utf8OverlongEncoding,
+        };
+        pub const Writer = std.io.Writer(*Self, Error, write);
+
+        const Self = @This();
+
+        pub fn write(self: *Self, bytes: []const u8) Error!usize {
+            const amt = try self.child_stream.write(bytes);
+            self.codepoints_written += try std.unicode.utf8CountCodepoints(bytes);
+            return amt;
```

Comment:
I don't think it's worth the complexity to handle, but there should at least be a comment here explaining that splitting a codepoint and writing it with two `writes` will fail.

Reply:
Good point. Do you think it makes sense to do `utf8CountCodepoints(bytes[0..amt])`, and disregard the `TruncatedInput` error?

Reply:
Didn't even think about partial writes from the `child_stream`. That makes this more complicated.

Calling `CodepointCountingWriter.write` with partial codepoints we can specify as unsupported (currently) since `clap` is the only user of this writer.

We have to handle partial writes from the `child_stream` though, since we have no control here. I'd recommend calling `self.child_stream.writeAll(bytes)` instead, to ensure all the bytes are written.

Reply:
I have to go to an appointment now, but I have an idea for how to make it behave right without too much hassle. Will try it when I get back!

Reply:
I added a modified version of the `utf8CountCodepoints` function, which returns the number of complete codepoints, if the string is truncated.

So `write`ing truncated UTF-8 will result in a partial write, leaving the incomplete codepoint for the next call.

This makes `writeAll` work as expected, and `write` works the same as before, reporting the same byte count as written to the child.

A bit more code then I would have liked, but conceptually simple